{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "9ee5105e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from glob import glob\n",
    "import os\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "7c005ade",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1.22.3'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy\n",
    "numpy.__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f926611e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from mito.genotyping import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "539f4957",
   "metadata": {},
   "source": [
    "## Load reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "5a956aef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to reference genome\n",
    "reference_path = '../../data/reference/mito_GRCh38_gimlet.fasta'\n",
    "\n",
    "with open(reference_path, 'r'):\n",
    "    reference_seq = next(SeqIO.parse(reference_path, \"fasta\"))\n",
    "reference = pd.Series(list(reference_seq.seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3e899e2e",
   "metadata": {},
   "source": [
    "## Load read matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "22cacc26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path to allele count files\n",
    "INPUT_PATH = '../../data/P9855_and_bulk_ac/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "c5e26b1b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading ../../data/P9855_and_bulk_ac/P3861_218.clean.dedup_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2085_S108_L004_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2089_S112_L004_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2090_S113_L004_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2091_S114_L004_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2093_S116_L005_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2096_S119_L005_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2101_S124_L006_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2102_S125_L006_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2104_S127_L006_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2110_S133_L007_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2111_S134_L007_ac.txt\n",
      "Reading ../../data/P9855_and_bulk_ac/P9855_2112_S135_L007_ac.txt\n"
     ]
    }
   ],
   "source": [
    "cell_count = []\n",
    "\n",
    "# Get all allele counts files in the input path\n",
    "cell_count_filenames = list(glob(os.path.join(INPUT_PATH, '*.txt'))) #*.txt\n",
    "\n",
    "for filename in sorted(cell_count_filenames):\n",
    "    print('Reading {}'.format(filename))\n",
    "    name = os.path.basename(filename).split('.')[0]\n",
    "    \n",
    "\n",
    "    # Load one allele count file\n",
    "    counts = pd.read_csv(filename, sep='\\t')\n",
    "    counts.name = name\n",
    "    cell_count.append(counts)\n",
    "    \n",
    "# Get sample order\n",
    "\n",
    "sample_list = []\n",
    "\n",
    "for filename in sorted(cell_count_filenames):\n",
    "    name = os.path.basename(filename).split('-')[0]\n",
    "    sample_list.append(name)\n",
    "\n",
    "sample_list_df = pd.DataFrame(sample_list)\n",
    "sample_list_df = sample_list_df.rename(columns={0: 'cell'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "accec7e2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a5b75ec0",
   "metadata": {},
   "outputs": [],
   "source": [
    "e_rates = [0.0001, 0.0002, 0.0003, 0.0004, 0.0005, 0.0006, 0.0007, 0.0008, 0.0009, 0.001]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac213a64",
   "metadata": {},
   "source": [
    "## Compute mutations probabilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "535f29c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cell_prob(e_rate):\n",
    "    \n",
    "    error_rate_when_no_mutation = error_rate_when_mutation = e_rate\n",
    "    p_mutation = 1 / 1000.0\n",
    "\n",
    "    cell_prob = []\n",
    "    for count in cell_count:\n",
    "        count = count.iloc[:reference.shape[0]]  # discard trailing positions\n",
    "        p = nucleotide_mutation_prob(\n",
    "            cell_counts=count,\n",
    "            reference=reference,\n",
    "            error_rate_when_no_mutation=error_rate_when_no_mutation,\n",
    "            error_rate_when_mutation=error_rate_when_mutation,\n",
    "            p_mutation=p_mutation,\n",
    "        )\n",
    "\n",
    "        cell_prob.append(p)\n",
    "        \n",
    "    # Compute P(mutation | read counts)\n",
    "    cells_p_mutation = []\n",
    "    for cell_prob in cell_prob:\n",
    "        p = mutation_prob(cell_prob, reference)\n",
    "        cells_p_mutation.append(p)\n",
    "        \n",
    "        \n",
    "    # Make mutation matrix\n",
    "    mutation_matrix = cells_p_mutation[0][['#CHR', 'POS']].copy()\n",
    "    mutation_matrix_data = pd.DataFrame(np.dstack([c['Prob_mutation'].values for c in cells_p_mutation]).squeeze())\n",
    "    mutation_matrix = pd.concat([mutation_matrix, mutation_matrix_data], axis=1)\n",
    "    return mutation_matrix      "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cfa83fce",
   "metadata": {},
   "source": [
    "## Filter pmat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "4d0bc336",
   "metadata": {},
   "outputs": [],
   "source": [
    "def flt_pmat(e_rate):\n",
    "    mutation_matrix = cell_prob(e_rate)\n",
    "\n",
    "    # Remove germline mutations\n",
    "    mutation_matrix = mutation_matrix.loc[mutation_matrix[0] < 0.9]\n",
    "    \n",
    "    # Probabilities for location with mutation at any of the cells\n",
    "    mutation_threshold = 0.9\n",
    "    data = mutation_matrix.set_index('POS').drop('#CHR', axis=1)\n",
    "    mutation_any_cell = data[(data > mutation_threshold).any(axis=1)]\n",
    "    \n",
    "    # select rows that have a value greater than mutation_threshold in more than x cells\n",
    "    mutation_threshold = 0.9\n",
    "    mask = (data.values > mutation_threshold).sum(axis=1) > 1 #create an array with boolean values and sum these along the axis and select rows with more than 1 True\n",
    "    two_cells_have_mut = data[mask]\n",
    "\n",
    "    # Fix matrix\n",
    "    half = len(two_cells_have_mut.columns) / 2\n",
    "\n",
    "    # Remove rows where 50% of cols are nan and save in new df\n",
    "    nan_rows = two_cells_have_mut.drop(two_cells_have_mut[(two_cells_have_mut.isna()).sum(axis=1)> half].index)\n",
    "\n",
    "    # Replace NaN with row mean\n",
    "    imputed = nan_rows.transpose().fillna(nan_rows.mean(axis=1)).transpose()\n",
    "    \n",
    "    # To skip removal of sites with overall high probability for mut\n",
    "    high_prob_rows = imputed.copy()\n",
    "    \n",
    "    # Replace 1.0\n",
    "    high_prob_rows = high_prob_rows.replace(1.0, 0.99999)\n",
    "    \n",
    "    np.set_printoptions(suppress=True)\n",
    "    e_rate_name = str(e_rate)\n",
    "    matrix_path = '../../data/P9855_matrix_output/' + e_rate_name + '.csv'\n",
    "    high_prob_rows.to_csv(matrix_path, index=False, sep=' ', header= False)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "c342cc49",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n",
      "/Users/johard/Documents/mcmt/MCMT/Notebooks/mito/genotyping.py:253: FutureWarning: The 'lookup' method is deprecated and will be removed in a future version. You can use DataFrame.melt and DataFrame.loc as a substitute.\n",
      "  prob_mutation['Prob_mutation'] = 1 - p.lookup(p.index, reference.values)\n"
     ]
    }
   ],
   "source": [
    "for e_rate in e_rates:\n",
    "    flt_pmat(e_rate)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c16fb41a",
   "metadata": {},
   "source": [
    "## Now run mt-SCITE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d893638",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "6492b2df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pmat_names</th>\n",
       "      <th>len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.0001</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0002</td>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0003</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0004</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0005</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.0006</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.0007</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>0.0008</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>0.0009</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.001</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  pmat_names  len\n",
       "0     0.0001  176\n",
       "1     0.0002   64\n",
       "2     0.0003   40\n",
       "3     0.0004   35\n",
       "4     0.0005   34\n",
       "5     0.0006   30\n",
       "6     0.0007   28\n",
       "7     0.0008   26\n",
       "8     0.0009   24\n",
       "9      0.001   24"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# prepare pmat data\n",
    "\n",
    "pmat_names = []\n",
    "shapes = []\n",
    "\n",
    "pmat_input_path = f'../../data/P9855_matrix_output/'\n",
    "#print(pmat_input_path)\n",
    "pmats = list(glob(os.path.join(pmat_input_path, '*.csv')))\n",
    "tree_name = []\n",
    "\n",
    "for filename in sorted(pmats):\n",
    "    name = os.path.basename(filename).split('-')[0].rsplit('.', 1)[0]\n",
    "    #print(name)\n",
    "    pmat_names.append(name)\n",
    "    df = pd.read_csv(filename, sep=' ', header=None)\n",
    "    shapes.append(len(df))\n",
    "\n",
    "# make df with pmat info\n",
    "pmat_data = pd.DataFrame(\n",
    "    {'pmat_names': pmat_names,\n",
    "     'len': shapes,\n",
    "    })\n",
    "\n",
    "pmat_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "c2f24199",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check n samples\n",
    "a_pmat = pd.read_csv('../../data/P9855_matrix_output/0.0006.csv', sep=' ', header=None)\n",
    "n_cells = a_pmat.shape[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "27e2142f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running tree inference for error rate 0.0001 repetition 1\n",
      "Running tree inference for error rate 0.0002 repetition 1\n",
      "Running tree inference for error rate 0.0003 repetition 1\n",
      "Running tree inference for error rate 0.0004 repetition 1\n",
      "Running tree inference for error rate 0.0005 repetition 1\n",
      "Running tree inference for error rate 0.0006 repetition 1\n",
      "Running tree inference for error rate 0.0007 repetition 1\n",
      "Running tree inference for error rate 0.0008 repetition 1\n",
      "Running tree inference for error rate 0.0009 repetition 1\n",
      "Running tree inference for error rate 0.001 repetition 1\n",
      "Running tree inference for error rate 0.0001 repetition 2\n",
      "Running tree inference for error rate 0.0002 repetition 2\n",
      "Running tree inference for error rate 0.0003 repetition 2\n",
      "Running tree inference for error rate 0.0004 repetition 2\n",
      "Running tree inference for error rate 0.0005 repetition 2\n",
      "Running tree inference for error rate 0.0006 repetition 2\n",
      "Running tree inference for error rate 0.0007 repetition 2\n",
      "Running tree inference for error rate 0.0008 repetition 2\n",
      "Running tree inference for error rate 0.0009 repetition 2\n",
      "Running tree inference for error rate 0.001 repetition 2\n",
      "Running tree inference for error rate 0.0001 repetition 3\n",
      "Running tree inference for error rate 0.0002 repetition 3\n",
      "Running tree inference for error rate 0.0003 repetition 3\n",
      "Running tree inference for error rate 0.0004 repetition 3\n",
      "Running tree inference for error rate 0.0005 repetition 3\n",
      "Running tree inference for error rate 0.0006 repetition 3\n",
      "Running tree inference for error rate 0.0007 repetition 3\n",
      "Running tree inference for error rate 0.0008 repetition 3\n",
      "Running tree inference for error rate 0.0009 repetition 3\n",
      "Running tree inference for error rate 0.001 repetition 3\n",
      "Running tree inference for error rate 0.0001 repetition 4\n",
      "Running tree inference for error rate 0.0002 repetition 4\n",
      "Running tree inference for error rate 0.0003 repetition 4\n",
      "Running tree inference for error rate 0.0004 repetition 4\n",
      "Running tree inference for error rate 0.0005 repetition 4\n",
      "Running tree inference for error rate 0.0006 repetition 4\n",
      "Running tree inference for error rate 0.0007 repetition 4\n",
      "Running tree inference for error rate 0.0008 repetition 4\n",
      "Running tree inference for error rate 0.0009 repetition 4\n",
      "Running tree inference for error rate 0.001 repetition 4\n",
      "Running tree inference for error rate 0.0001 repetition 5\n",
      "Running tree inference for error rate 0.0002 repetition 5\n",
      "Running tree inference for error rate 0.0003 repetition 5\n",
      "Running tree inference for error rate 0.0004 repetition 5\n",
      "Running tree inference for error rate 0.0005 repetition 5\n",
      "Running tree inference for error rate 0.0006 repetition 5\n",
      "Running tree inference for error rate 0.0007 repetition 5\n",
      "Running tree inference for error rate 0.0008 repetition 5\n",
      "Running tree inference for error rate 0.0009 repetition 5\n",
      "Running tree inference for error rate 0.001 repetition 5\n",
      "Running tree inference for error rate 0.0001 repetition 6\n",
      "Running tree inference for error rate 0.0002 repetition 6\n",
      "Running tree inference for error rate 0.0003 repetition 6\n",
      "Running tree inference for error rate 0.0004 repetition 6\n",
      "Running tree inference for error rate 0.0005 repetition 6\n",
      "Running tree inference for error rate 0.0006 repetition 6\n",
      "Running tree inference for error rate 0.0007 repetition 6\n",
      "Running tree inference for error rate 0.0008 repetition 6\n",
      "Running tree inference for error rate 0.0009 repetition 6\n",
      "Running tree inference for error rate 0.001 repetition 6\n",
      "Running tree inference for error rate 0.0001 repetition 7\n",
      "Running tree inference for error rate 0.0002 repetition 7\n",
      "Running tree inference for error rate 0.0003 repetition 7\n",
      "Running tree inference for error rate 0.0004 repetition 7\n",
      "Running tree inference for error rate 0.0005 repetition 7\n",
      "Running tree inference for error rate 0.0006 repetition 7\n",
      "Running tree inference for error rate 0.0007 repetition 7\n",
      "Running tree inference for error rate 0.0008 repetition 7\n",
      "Running tree inference for error rate 0.0009 repetition 7\n",
      "Running tree inference for error rate 0.001 repetition 7\n",
      "Running tree inference for error rate 0.0001 repetition 8\n",
      "Running tree inference for error rate 0.0002 repetition 8\n",
      "Running tree inference for error rate 0.0003 repetition 8\n",
      "Running tree inference for error rate 0.0004 repetition 8\n",
      "Running tree inference for error rate 0.0005 repetition 8\n",
      "Running tree inference for error rate 0.0006 repetition 8\n",
      "Running tree inference for error rate 0.0007 repetition 8\n",
      "Running tree inference for error rate 0.0008 repetition 8\n",
      "Running tree inference for error rate 0.0009 repetition 8\n",
      "Running tree inference for error rate 0.001 repetition 8\n",
      "Running tree inference for error rate 0.0001 repetition 9\n",
      "Running tree inference for error rate 0.0002 repetition 9\n",
      "Running tree inference for error rate 0.0003 repetition 9\n",
      "Running tree inference for error rate 0.0004 repetition 9\n",
      "Running tree inference for error rate 0.0005 repetition 9\n",
      "Running tree inference for error rate 0.0006 repetition 9\n",
      "Running tree inference for error rate 0.0007 repetition 9\n",
      "Running tree inference for error rate 0.0008 repetition 9\n",
      "Running tree inference for error rate 0.0009 repetition 9\n",
      "Running tree inference for error rate 0.001 repetition 9\n",
      "Running tree inference for error rate 0.0001 repetition 10\n",
      "Running tree inference for error rate 0.0002 repetition 10\n",
      "Running tree inference for error rate 0.0003 repetition 10\n",
      "Running tree inference for error rate 0.0004 repetition 10\n",
      "Running tree inference for error rate 0.0005 repetition 10\n",
      "Running tree inference for error rate 0.0006 repetition 10\n",
      "Running tree inference for error rate 0.0007 repetition 10\n",
      "Running tree inference for error rate 0.0008 repetition 10\n",
      "Running tree inference for error rate 0.0009 repetition 10\n",
      "Running tree inference for error rate 0.001 repetition 10\n"
     ]
    }
   ],
   "source": [
    "# Run mt-SCITE\n",
    "\n",
    "SCITE_PATH = '../../../mt-SCITE'\n",
    "PMAT_PATH = f'../../data/P9855_matrix_output/'\n",
    "OUTPUT = f'../../../mt-SCITE/mt-SCITE_output/P9855/'\n",
    "\n",
    "for rep in range(1,11): #11\n",
    "    \n",
    "    for pmat in pmat_names:        \n",
    "        run_id = pmat + '_' + str(rep)\n",
    "        print('Running tree inference for error rate ' + pmat + ' repetition ' + str(rep))\n",
    "\n",
    "        # Get number of mutations\n",
    "        n = pmat_data.loc[pmat_data['pmat_names'] == pmat, 'len'].iloc[0].astype(str)\n",
    "\n",
    "        try:\n",
    "            os.makedirs(OUTPUT + '/stdout/') \n",
    "        except FileExistsError :\n",
    "            pass\n",
    "        except :\n",
    "            raise\n",
    "\n",
    "        ! $SCITE_PATH/mt-SCITE/scite -i $PMAT_PATH/$pmat\\.csv -n $n -m $n_cells -r 1 -l 200000 -fd 0.0001 -ad 0.0001 -cc 0.0 -s -a -o $OUTPUT/$run_id 1> $OUTPUT/stdout/$run_id\\.stdout.txt\n",
    "        \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "270e8a3b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mtSCITE",
   "language": "python",
   "name": "mtscite"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
